### PR TITLE
[Reland] Detect against invalid variant index for LibStdC++ std::variant data formatters

### PR DESCRIPTION
**Pull request by jeffreytan81**
This is relanding of https://github.com/llvm/llvm-project/pull/69253. `TestTemplatePackArgs.py` is passing now.

https://github.com/llvm/llvm-project/pull/68012/files added new data formatters for LibStdC++ std::variant.

However, this formatter can crash if std::variant's index field has invalid value (exceeds the number of template arguments).
This can happen if the current IP stops at a place std::variant is not initialized yet.

This patch fixes the crash by ensuring the index is a valid value and fix GetNthTemplateArgument() to make sure it is not crashing.

### CODE REVIEW

File: lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp [hunk 1]
```diff
@@ -7183,7 +7183,8 @@ GetNthTemplateArgument(const clang::ClassTemplateSpecializationDecl *decl,
   // (including the ones preceding the parameter pack).
   const auto &pack = args[last_idx];
   const size_t pack_idx = idx - last_idx;
-  assert(pack_idx < pack.pack_size() && "parameter pack index out-of-bounds");
+  if (pack_idx >= pack.pack_size())
+    return nullptr;
```
Comment (DavidSpickett) [inline lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:7187]: Is this change needed now that the formatter is checking the number of arguments up front?

Comment (bulbazord) [reply lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:7187]: +1, we should still include this change. We shouldn't crash if the `idx` argument is too high.

**Review by clayborg — APPROVED** [review]
